rfcs/0125-bootspec.md

```diff
@@ -143,18 +145,31 @@ Using the following JSON:
     // Top level path of the closure, in case some spelunking is required
     "toplevel": "/nix/store/xxx-nixos-system-xxx",
 
+    // Specialisations are optional
     "specialisation": {
       // <name> corresponds to <name> in specialisation.<name>.configuration.
       // Note that it is not valid for a specialisation to be a different version than the main document.
       "<name>": {
         // A full Bootspec v1 document, _without_ the `"v1"` key.
         // Note that it is not valid for a specialisation to have further specialisations.
       }
+    },
+
+    // Optional user-specific metadata
+    "extensions": {
+      // <namespace> corresponds to a user-provided namespace
+      // to reduce the chances of collision when using
+      // multiple extensions at once
+      "<namespace>": {
+        // Any kind of representable JSON is valid here
+      }
     }
   }
 }
 ```
 
+An *optional* field mean: either missing or present, **not null**.
```

Comment:
We should go back then to the RFC-0125 PR discussion because this is something we decided during a meeting to disambiguate what was meant by "optional".

I agree with you that we should not spec what happens in the custom data, this is not the role of the spec and consumers can do whatever they want in the `HashMap<String, serde_json::Value>` implementation at the moment.

I don't necessarily agree that `extensions: null` should be a valid representation of `extensions: HashMap::new()`.

In JavaScript, `null` and `undefined` are two different values for `!==` (though they are equal for `==`), the idea behind constraining the values is to make it easier to reason in other languages such as those.

Making so that `null` never happen in the document is a nice way to eliminate this class of bugs IMHO.

Reply:
> Making so that `null` never happen in the document is a nice way to eliminate this class of bugs IMHO.

You've convinced me with that single line; I think it's fine to deny `null`. :smile: